File: data_pipeline/validate-pilot.mjs
import { createHash } from "node:crypto";
import { readFile } from "node:fs/promises";

const allowedLicenses = new Set(["MIT", "BSD-2-Clause", "BSD-3-Clause", "Apache-2.0"]);
const repositorySchema = JSON.parse(await readFile(new URL("./manifests/repository.schema.json", import.meta.url), "utf8"));
const manifests = await readJsonl(new URL("./manifests/pilot-repositories.jsonl", import.meta.url));
const accepted = await readJsonl(new URL("./samples/accepted.jsonl", import.meta.url));
const rejected = await readJsonl(new URL("./samples/rejected.jsonl", import.meta.url));

assert(manifests.length === 5, `expected 5 repository manifests, found ${manifests.length}`);
assert(accepted.length + rejected.length === 8, "expected 8 extracted code units");
assert(
  JSON.stringify(repositorySchema.properties?.license?.properties?.spdx_id?.enum) === JSON.stringify([...allowedLicenses]),
  "repository schema license allowlist differs from validator policy",
);

const manifestByRepo = new Map();
for (const manifest of manifests) {
  assert(manifest.schema_version === "0.1", "unsupported manifest schema version");
  assert(typeof manifest.repository_id === "string" && manifest.repository_id.includes("/"), "invalid repository_id");
  assert(/^https:\/\/github\.com\//.test(manifest.url), `invalid repository URL for ${manifest.repository_id}`);
  assert(/^[0-9a-f]{40}$/.test(manifest.commit_sha), `invalid commit for ${manifest.repository_id}`);
  assert(allowedLicenses.has(manifest.license?.spdx_id), `license rejected for ${manifest.repository_id}`);
  assert(manifest.license?.verified === true, `license not verified for ${manifest.repository_id}`);
  assert(manifest.build?.exit_code === 0 && String(manifest.build.status).startsWith("passed"), `build not passed for ${manifest.repository_id}`);
  assert(manifest.duplicate?.is_fork === false, `fork rejected for ${manifest.repository_id}`);
  assert(!manifestByRepo.has(manifest.repository_id), `duplicate repository ${manifest.repository_id}`);
  manifestByRepo.set(manifest.repository_id, manifest);
}

const ids = new Set();
const hashes = new Set();
for (const sample of [...accepted, ...rejected]) {
  assert(!ids.has(sample.sample_id), `duplicate sample id ${sample.sample_id}`);
  ids.add(sample.sample_id);
  assert(manifestByRepo.has(sample.metadata?.source_repo), `unknown source repository for ${sample.sample_id}`);
  assert(manifestByRepo.get(sample.metadata.source_repo).commit_sha === sample.metadata.source_commit, `commit mismatch for ${sample.sample_id}`);
  assert(["G1", "G2"].includes(sample.metadata?.granularity), `pilot granularity rejected for ${sample.sample_id}`);
  assert(sample.messages?.length === 2 && sample.messages[0]?.role === "user" && sample.messages[1]?.role === "assistant", `invalid messages for ${sample.sample_id}`);
  assert(sample.messages[0].content.trim().length > 0 && sample.messages[1].content.trim().length > 0, `empty instruction or target for ${sample.sample_id}`);
  const digest = createHash("sha256").update(sample.messages[1].content.replaceAll("\r\n", "\n").trim()).digest("hex");
  assert(digest === sample.target?.content_sha256, `target hash mismatch for ${sample.sample_id}`);
  assert(!hashes.has(digest), `exact duplicate target ${sample.sample_id}`);
  hashes.add(digest);
}

for (const sample of accepted) {
  assert(sample.metadata.quality?.alignment === "pass", `alignment failed for accepted sample ${sample.sample_id}`);
  assert(sample.metadata.quality?.granularity_match === "pass", `granularity failed for accepted sample ${sample.sample_id}`);
  assert(sample.metadata.quality?.solvability === "pass", `solvability failed for accepted sample ${sample.sample_id}`);
}

for (const sample of rejected) {
  assert(sample.metadata?.status === "rejected", `missing rejected status for ${sample.sample_id}`);
  assert(Array.isArray(sample.metadata?.rejection_reasons) && sample.metadata.rejection_reasons.length > 0, `missing rejection reason for ${sample.sample_id}`);
}

console.log(JSON.stringify({ repositories: manifests.length, extractedUnits: accepted.length + rejected.length, acceptedSamples: accepted.length, rejectedSamples: rejected.length, exactDuplicateTargets: 0, passed: true }));

async function readJsonl(url) {
  const content = await readFile(url, "utf8");
  return content.split(/\r?\n/).filter((line) => line.trim().length > 0).map((line, index) => {
    try {
      return JSON.parse(line);
    } catch (error) {
      throw new Error(`${url.pathname}:${index + 1}: ${error instanceof Error ? error.message : "invalid JSON"}`);
    }
  });
}

function assert(condition, message) {
  if (!condition) throw new Error(message);
}

File: data_pipeline/validate_pilot.py
from __future__ import annotations

import hashlib
import json
from pathlib import Path
from typing import Any


ROOT = Path(__file__).resolve().parent
ALLOWED = ["MIT", "BSD-2-Clause", "BSD-3-Clause", "Apache-2.0"]


def rows(path: Path) -> list[dict[str, Any]]:
    return [json.loads(line) for line in path.read_text(encoding="utf-8").splitlines() if line.strip()]


def require(condition: bool, message: str) -> None:
    if not condition:
        raise ValueError(message)


def main() -> None:
    schema = json.loads((ROOT / "manifests/repository.schema.json").read_text())
    manifests, accepted, rejected = rows(ROOT / "manifests/pilot-repositories.jsonl"), rows(ROOT / "samples/accepted.jsonl"), rows(ROOT / "samples/rejected.jsonl")
    require(len(manifests) == 5, f"expected 5 repository manifests, found {len(manifests)}")
    require(len(accepted) + len(rejected) == 8, "expected 8 extracted code units")
    require(schema["properties"]["license"]["properties"]["spdx_id"]["enum"] == ALLOWED, "repository schema license allowlist differs from validator policy")
    by_repo: dict[str, Any] = {}
    for manifest in manifests:
        repo = manifest["repository_id"]
        require(manifest["schema_version"] == "0.1" and "/" in repo, "invalid manifest")
        require(manifest["url"].startswith("https://github.com/"), f"invalid repository URL for {repo}")
        require(bool(__import__("re").fullmatch(r"[0-9a-f]{40}", manifest["commit_sha"])), f"invalid commit for {repo}")
        require(manifest["license"]["spdx_id"] in ALLOWED and manifest["license"]["verified"] is True, f"license rejected for {repo}")
        require(manifest["build"]["exit_code"] == 0 and str(manifest["build"]["status"]).startswith("passed"), f"build not passed for {repo}")
        require(manifest["duplicate"]["is_fork"] is False and repo not in by_repo, f"duplicate/fork repository {repo}")
        by_repo[repo] = manifest
    ids: set[str] = set(); hashes: set[str] = set()
    for sample in accepted + rejected:
        sample_id, metadata = sample["sample_id"], sample["metadata"]
        require(sample_id not in ids, f"duplicate sample id {sample_id}"); ids.add(sample_id)
        require(metadata["source_repo"] in by_repo and by_repo[metadata["source_repo"]]["commit_sha"] == metadata["source_commit"], f"provenance mismatch for {sample_id}")
        require(metadata["granularity"] in {"G1", "G2"}, f"pilot granularity rejected for {sample_id}")
        messages = sample["messages"]
        require(len(messages) == 2 and [row["role"] for row in messages] == ["user", "assistant"], f"invalid messages for {sample_id}")
        target = messages[1]["content"].replace("\r\n", "\n").strip()
        digest = hashlib.sha256(target.encode()).hexdigest()
        require(digest == sample["target"]["content_sha256"] and digest not in hashes, f"target hash/duplicate mismatch for {sample_id}"); hashes.add(digest)
    for sample in accepted:
        quality = sample["metadata"]["quality"]
        require(all(quality[key] == "pass" for key in ("alignment", "granularity_match", "solvability")), f"quality failed for {sample['sample_id']}")
    for sample in rejected:
        require(sample["metadata"].get("status") == "rejected" and sample["metadata"].get("rejection_reasons"), f"missing rejection reason for {sample['sample_id']}")
    print(json.dumps({"repositories": len(manifests), "extractedUnits": len(accepted) + len(rejected), "acceptedSamples": len(accepted), "rejectedSamples": len(rejected), "exactDuplicateTargets": 0, "passed": True}))


if __name__ == "__main__":
    main()
